### PR TITLE
MOTECH-2298: Changed the way BaseListeners retrieve application context [master]

### PR DESCRIPTION
**Pull request by ngraczewski**
MDS BaseListeners will now try to retrieve entities bundle context from the bundle context before waiting for the notification.

### CODE REVIEW

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 1]
```diff
@@ -24,6 +31,9 @@
  */
 public abstract class BaseListener<T> implements ServiceListener {
 
+    private static final String ENTITIES_BUNDLE_SYMBOLIC_NAME = "org.motechproject.motech-platform-dataservices-entities";
```
Comment (sebbrudzinski) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: Constants.BundleNames.MDS_ENTITIES_SYMBOLIC_NAME

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 2]
```diff
@@ -63,14 +73,77 @@ public void serviceChanged(ServiceEvent event) {
...
+        try {
+            entitiesBundleApplicationContext = getEntitiesBundleApplicationContext(getEntitiesBundle());
+        } catch (InvalidSyntaxException e) {
+            logger.error("Invalid syntax of the filter passed to the bundle context");
+        }
+
+        return entitiesBundleApplicationContext;
+    }
+
+    private Bundle getEntitiesBundle() {
```
Comment (sebbrudzinski) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: Can we reuse util methods from MdsBundleHelper?

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 3]
```diff
@@ -24,6 +32,9 @@
  */
 public abstract class BaseListener<T> implements ServiceListener {
 
+    private static final String MDS_ENTITIES_SYMBOLIC_NAME = "org.motechproject.motech-platform-dataservices-entities";
```
Comment (sebbrudzinski) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: This constant already exists in the Constants class. I've mentioned this in the previous comment.

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 4]
```diff
@@ -63,14 +74,68 @@ public void serviceChanged(ServiceEvent event) {
...
+
+            BundleContext entitiesbundleContext = entitiesBundle.getBundleContext();
+            List<ServiceReference<?>> references = getApplicationContextReferences(entitiesbundleContext);
+
+            if (references.size() == 1) {
+                context = (ApplicationContext) entitiesbundleContext.getService(references.get(0));
+            } else if (references.size() > 1) {
+                logger.warn("Multiple entities bundle application contexts found");
+            } else {
+                logger.debug("No entities bundle application context found");
```
Comment (pgesek) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: shouldn't this be a warn as well?

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 5]
```diff
@@ -63,14 +74,68 @@ public void serviceChanged(ServiceEvent event) {
...
+    }
+
+    private List<ServiceReference<?>> getApplicationContextReferences(BundleContext bundleContext)
+            throws InvalidSyntaxException {
+
+        List<ServiceReference<?>> references = new ArrayList<>();
+
+        if (bundleContext != null) {
+            references.addAll(bundleContext.getServiceReferences(
+                    ConfigurableApplicationContext.class,
```
Comment (pgesek) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: why ConfigurableApplicationContext, not ApplicationContext?

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java [hunk 6]
```diff
@@ -63,14 +74,68 @@ public void serviceChanged(ServiceEvent event) {
...
+            logger.debug("Entities bundle not found");
+        }
+
+        return context;
+    }
+
+    private List<ServiceReference<?>> getApplicationContextReferences(BundleContext bundleContext)
+            throws InvalidSyntaxException {
+
+        List<ServiceReference<?>> references = new ArrayList<>();
```
Comment (pgesek) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: why ?, can't we be type safe here?

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java (hunk 5 above)
Comment (ngraczewski) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: Changed to ApplicationContext.

File: platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java (hunk 6 above)
Comment (ngraczewski) [inline platform/mds/mds/src/main/java/org/motechproject/mds/listener/records/BaseListener.java]: Made type safe by changing to ApplicationContext.

**Pull request merged**